From c738e38a5ad699e994a618c89d55420b391d90e1 Mon Sep 17 00:00:00 2001
From: Senthilnathan Veppur <senthilnathanx.veppur@intel.com>
Date: Mon, 30 Nov 2015 18:28:37 +0530
Subject: [PATCH 1793/2508] ASoC: Intel: Skylake: Release topology firmware
 memory

commit 97fa61240440761e9e9d8b5c72013af6c2eb6f45 from
https://github.com/01org/linux-apollolake-i

Add topology fw variable to platform structure.
Release allocated firmware topology memory in remove function
of platform.

Change-Id: I7d1f64d1286fa0ff2e44832ccd41d38919f7866a
Signed-off-by: Senthilnathan Veppur <senthilnathanx.veppur@intel.com>
Reviewed-on: https://git-gar-1.devtools.intel.com/gerrit/7633
Reviewed-by: Babu, Ramesh <ramesh.babu@intel.com>
Tested-by: Babu, Ramesh <ramesh.babu@intel.com>
---
 sound/soc/intel/skylake/skl-pcm.c      |   14 ++++++++++++++
 sound/soc/intel/skylake/skl-topology.c |    7 +++----
 sound/soc/intel/skylake/skl.h          |    2 ++
 3 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/sound/soc/intel/skylake/skl-pcm.c b/sound/soc/intel/skylake/skl-pcm.c
index 96516f0..5f51fe1 100644
--- a/sound/soc/intel/skylake/skl-pcm.c
+++ b/sound/soc/intel/skylake/skl-pcm.c
@@ -21,6 +21,7 @@
 
 #include <linux/pci.h>
 #include <linux/pm_runtime.h>
+#include <linux/firmware.h>
 #include <sound/pcm_params.h>
 #include <sound/soc.h>
 #include "skl.h"
@@ -1065,8 +1066,21 @@ static int skl_platform_soc_probe(struct snd_soc_platform *platform)
 
 	return 0;
 }
+
+static int skl_platform_soc_remove(struct snd_soc_platform *platform)
+{
+	struct hdac_ext_bus *ebus = dev_get_drvdata(platform->dev);
+	struct skl *skl = ebus_to_skl(ebus);
+
+	if (skl->fw) {
+		release_firmware(skl->fw);
+		skl->fw = NULL;
+	}
+	return 0;
+}
 static struct snd_soc_platform_driver skl_platform_drv  = {
 	.probe		= skl_platform_soc_probe,
+	.remove         = skl_platform_soc_remove,
 	.ops		= &skl_platform_ops,
 	.pcm_new	= skl_pcm_new,
 	.pcm_free	= skl_pcm_free,
diff --git a/sound/soc/intel/skylake/skl-topology.c b/sound/soc/intel/skylake/skl-topology.c
index c070021..db8d0bb 100644
--- a/sound/soc/intel/skylake/skl-topology.c
+++ b/sound/soc/intel/skylake/skl-topology.c
@@ -1497,11 +1497,10 @@ static struct snd_soc_tplg_ops skl_tplg_ops  = {
 int skl_tplg_init(struct snd_soc_platform *platform, struct hdac_ext_bus *ebus)
 {
 	int ret;
-	const struct firmware *fw;
 	struct hdac_bus *bus = ebus_to_hbus(ebus);
 	struct skl *skl = ebus_to_skl(ebus);
 
-	ret = request_firmware(&fw, "dfw_sst.bin", bus->dev);
+	ret = request_firmware(&skl->fw, "dfw_sst.bin", bus->dev);
 	if (ret < 0) {
 		dev_err(bus->dev, "tplg fw %s load failed with %d\n",
 				"dfw_sst.bin", ret);
@@ -1513,10 +1512,10 @@ int skl_tplg_init(struct snd_soc_platform *platform, struct hdac_ext_bus *ebus)
 	 * any other index
 	 */
 	ret = snd_soc_tplg_component_load(&platform->component,
-					&skl_tplg_ops, fw, 0);
+					&skl_tplg_ops, skl->fw, 0);
 	if (ret < 0) {
 		dev_err(bus->dev, "tplg component load failed%d\n", ret);
-		release_firmware(fw);
+		release_firmware(skl->fw);
 		return -EINVAL;
 	}
 
diff --git a/sound/soc/intel/skylake/skl.h b/sound/soc/intel/skylake/skl.h
index c43e0f9..999623f 100644
--- a/sound/soc/intel/skylake/skl.h
+++ b/sound/soc/intel/skylake/skl.h
@@ -24,6 +24,7 @@
 #include <sound/hda_register.h>
 #include <sound/hdaudio_ext.h>
 #include "skl-nhlt.h"
+#include "skl-tplg-interface.h"
 
 #define SKL_SUSPEND_DELAY 2000
 
@@ -70,6 +71,7 @@ struct skl {
 
 	struct skl_dsp_resource resource;
 	struct list_head ppl_list;
+	const struct firmware *fw;
 
 	struct skl_debug *debugfs;
 	bool nhlt_override;
-- 
1.7.5.4

